concordium-consensus/src/Concordium/GlobalState/Basic/BlockState/Modules.hs

```diff
@@ -37,24 +42,57 @@ type ModuleIndex = Word64
 --------------------------------------------------------------------------------
 
 -- |A module contains both the module interface and the raw source code of the
--- module.
-data Module = Module {
+-- module. The module is parameterized by the wasm version, which determines the shape
+-- of the module interface.
+data ModuleV v = ModuleV {
   -- | The instrumented module, ready to be instantiated.
-  interface :: !ModuleInterface,
+  moduleVInterface :: !(GSWasm.ModuleInterfaceV v),
   -- | The raw module binary source.
-  source :: !WasmModule
+  moduleVSource :: !(WasmModuleV v)
   } deriving (Show)
 
+-- |A module, either of version 0 or 1. This is only used when storing a module
+-- independently, e.g., in the module table. When a module is referenced from a
+-- contract instance we use the ModuleV type directly so we may tie the version
+-- of the module to the version of the instance.
+data Module where
+  ModuleV0 :: ModuleV GSWasm.V0 -> Module
+  ModuleV1 :: ModuleV GSWasm.V1 -> Module
+  deriving(Show)
+
+-- |Helper (internal to the module) to convert from a module to an interface.
```

Comment:
Not a big thing, but not sure you need to document that explicit if it's internal or not when it's not exposed by the module. 
For example `fromModule` is also internal but does not have this remark. Either it should be stated explicitly in the comment of all the (internal) functions  or none at all, inconsistency can be a cause of confusion.

Reply:
It was to make the intent clearer. I inlined this now.